bitcoin/src/network/constants.rs

```diff
@@ -92,22 +90,115 @@ impl Network {
     /// # Examples
     ///
     /// ```rust
-    /// use bitcoin::network::constants::Network;
+    /// use bitcoin::network::constants::{Network, Magic};
     ///
     /// let network = Network::Bitcoin;
-    /// assert_eq!(network.magic(), 0xD9B4BEF9);
+    /// assert_eq!(network.magic(), Magic::from_bytes([0xF9, 0xBE, 0xB4, 0xD9]));
     /// ```
-    pub fn magic(self) -> u32 {
-        // Note: any new entries here must be added to `from_magic` above
+    pub fn magic(self) -> Magic {
+        Magic::from(self)
+    }
+}
+
+/// Network magic bytes to identify the cryptocurrency network the message was intended for.
+#[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+pub struct Magic([u8; 4]);
+
+impl Magic {
+    /// Create network magic from bytes.
+    pub fn from_bytes(bytes: [u8; 4]) -> Magic {
+        Magic(bytes)
+    }
+
+    /// Get network magic bytes.
+    pub fn to_bytes(self) -> [u8; 4] {
+        self.0
+    }
```

Comment:
Perhaps we should have `as_bytes` as well? @Kixunil would be that useful?

Reply:
`AsRef<[u8; 4]>` and `AsRef<[u8]>` might make sense if we want to commit to the internal representation. (Likely OK)

Reply:
`as_bytes()` sounds more correct than `to_bytes()` since the magic is just a set of bytes. No conversion _to_ bytes is happening.

Similar to how `str` has `as_bytes()` where as `u32` has `to_le_bytes()`/`to_be_bytes()`.

Reply:
Also added `AsRef<[u8; 4]>` and `AsRef<[u8]>` for `Magic`.

Reply:
That's not how Rust naming works. `as_*` refers to converting a reference to a different reference which is not happening here. `to_bytes()` was correct naming.

Reply:
The doc is: https://rust-lang.github.io/api-guidelines/naming.html (Ctl-f "into" for the section) - read, re-read, bookmark, read again, cry a little bit, repeat, seemingly forever for these names `to`, `into`, `as` :)

Reply:
Sadly, that part of the guidelines is not even well written. We had to figure out from `std` API.